File: c_solutions_31-40/Euler_31.c
#include "euler_util.h"

#define CAP 8

const int currency[CAP] = {1, 2, 5, 10, 20, 50, 100, 200};
int total[CAP] =          {0, 0, 0,  0,  0,  0,   0,   0};

const int limit = 200;


static inline int calculate_total(void)
{
    int sum = 0;

    for (int i=0; i < CAP; i++) 
        sum += total[i];

    return sum;
}

int main(int argc, char *argv[])
{
    float start = timeit();

    int ways = 0;
    int idx = 0;

    while (total[idx] <= limit && (idx != CAP)) {

        total[idx] += currency[idx];
        int sum = calculate_total();

        if (sum < limit)
            idx = 0;
        else {
            if (sum == limit)
                ways++;
            total[idx] = 0;
            idx++;
        } 
    }

    float stop = timeit();

    printf("Answer: %d\n", ways);
    printf("Time: %f\n", stop - start);

    return 0;
}

File: py_solutions_31-40/Euler_31.py
# -*- coding: utf-8 -*-
'''
In England the currency is made up of pound, £, and pence, p, and there are eight coins in general circulation:

    1p, 2p, 5p, 10p, 20p, 50p, £1 (100p) and £2 (200p).

It is possible to make £2 in the following way:

    1×£1 + 1×50p + 2×20p + 1×5p + 1×2p + 3×1p

How many different ways can £2 be made using any number of coins?
'''

import timeit

start = timeit.default_timer()


try:
    range = xrange
except NameError:
    pass


def euler_31():
    limit = 200
    currency = [1, 2, 5, 10, 20, 50, 100, 200]
    no_currency = len(currency)
    amount = [0] * no_currency
    idx = ways = 0
    while idx < no_currency:
        amount[idx] += currency[idx]
        total = sum(amount)
        if total < limit:
            idx = 0
        else:
            amount[idx] = 0
            if total == limit: ways += 1
            idx += 1
    return ways

print('Answer: {}'.format(euler_31()))
stop = timeit.default_timer()
print('Time: {0:9.5}'.format(stop - start))

